### PR TITLE
fix: bank reconciliation rounding preventing clearance

### LINKED ISSUE #35440: v14 Bank Reconciliation - does not clear all when reconciling multiple vouchers with a single Bank Transaction in a single step
### Information about bug

When there are multiple vouchers offered to match a single Bank Transaction, and the sum of the vouchers will clear the Bank Transaction, not all (and it appears only one) voucher is marked cleared, yet both are added to and reconciled against the Bank Transaction.

Screen Shot 2023-05-27 at 15 47 52

All vouchers need to be cleared.

Workaround: Reconcile each voucher against the Bank Transaction separately.

### Module

accounts

### Version

ERPNext >= v14.19

### Installation method

None

### Relevant log output / Stack trace / Full Error Message.

Introduced: #34476

### PR DESCRIPTION
**Pull request by casesolved-co-uk**
Fixes #35440 

version-14-hotfix

### CODE REVIEW

File: erpnext/accounts/doctype/bank_transaction/bank_transaction.py [hunk 1]
```diff
@@ -8,8 +8,23 @@
...
+
+	def flt(self, value):
+		"frappe needs a money datatype"
+		if self._cy_precision is None:
+			import math
+
+			fraction_units = (
+				frappe.db.get_value("Currency", self.currency, "fraction_units", cache=True) or 100
+			)
+			self._cy_precision = math.log(fraction_units, 10)
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_transaction/bank_transaction.py]: `log(1000, 10) != 3`
```suggestion
			self._cy_precision = round(math.log(fraction_units, 10))
```

`math.log` returns a float, while `flt` expects an integer as precision.

File: erpnext/accounts/doctype/bank_transaction/bank_transaction.py [hunk 2]
```diff
@@ -8,8 +8,23 @@
 
 
 class BankTransaction(StatusUpdater):
+	_cy_precision = None
+
+	def flt(self, value):
+		"frappe needs a money datatype"
+		if self._cy_precision is None:
+			import math
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_transaction/bank_transaction.py]: This import should be at the top of the file

File: erpnext/accounts/doctype/bank_transaction/bank_transaction.py [hunk 3]
```diff
@@ -8,8 +8,23 @@
 
 
 class BankTransaction(StatusUpdater):
+	_cy_precision = None
+
+	def flt(self, value):
+		"frappe needs a money datatype"
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_transaction/bank_transaction.py]:
```suggestion
```

Oka, but this is not helpful as a docstring

File: erpnext/accounts/doctype/bank_transaction/bank_transaction.py [hunk 4]
```diff
@@ -8,8 +8,23 @@
 
 
 class BankTransaction(StatusUpdater):
+	_cy_precision = None
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_transaction/bank_transaction.py]:
```suggestion
	_currency_precision = None
```

I don't like guessing

**Review by barredterra — CHANGES_REQUESTED** [review]
Not sure about the general implementation approach. How is this handled in other doctypes?

**Comment by casesolved-co-uk** [comment]
> Not sure about the general implementation approach. How is this handled in other doctypes?

You're the reviewer, you tell me

**Comment by barredterra** [comment]
> You're the reviewer, you tell me

Nope. You are proposing this change, so if you want to get it merged you'll have to do the work and convince the reviewers that it is correct.

**Comment by casesolved-co-uk** [comment]
> > You're the reviewer, you tell me
> 
> Nope. You are proposing this change, so if you want to get it merged you'll have to do the work and convince the reviewers that it is correct.

It's a bug fix, and it's been tested to fix the bug. Any architectural policies are your preprogative. Anyone would think frappe like to have bugs in their code given the resistance.

**Comment by barredterra** [comment]
Anyone would think you're just here to ~~display your arrogance instead of collaborating productively~~ delight everybody with your important contributions and graceful comments.

**Comment by casesolved-co-uk** [comment]
> Anyone would think you're just here to display your arrogance instead of collaborating productively.

On the contrary, I think it's arrogant to expect everyone fixing bugs in your software to know every single detail of the software architecture.

No wonder there are thousands of ignored issues and pull requests.

Explain to me how fixing bugs in your software and amending your review comments is arrogant?

**Comment by barredterra** [comment]
It's not my software. I'm reviewing you PR on my own time and money. If you're not willing to be friendly and put in the required effort, I will simply stop reviewing your PRs.

**Comment by casesolved-co-uk** [comment]
> It's not my software. I'm reviewing you PR on my own time and money. If you're not willing to be friendly and put in the required effort, I will simply stop reviewing your PRs.

I have been friendly, it's you calling me names.
I have put in the required effort by finding and fixing the bug and testing it.
Your review is almost entirely cosmetic and personal preference and yet I STILL made the changes.

If it's not your software, why are you a maintainer and reviewer authorized by frappe?
How is calling me arrogant friendly?
How is fixing bugs in your software and amending your review comments not "collaborating productively"?
How is not recognising the value and contributions of mine not arrogant?

We are all contributing on our own time and money given this is open source software. Where is your appreciation?

I refer you to your own code of conduct:

> Examples of behavior that contributes to creating a positive environment include:
> 
> Using welcoming and inclusive language
> Being respectful of differing viewpoints and experiences
> Gracefully accepting constructive criticism
> Focusing on what is best for the community
> Showing empathy towards other community members
>
> Examples of unacceptable behavior by participants include:
> 
> The use of sexualized language or imagery and unwelcome sexual attention or advances
> Trolling, insulting/derogatory comments, and personal or political attacks
> Public or private harassment
> Publishing others' private information, such as a physical or electronic address, without explicit permission
> Other conduct which could reasonably be considered inappropriate in a professional setting

**Comment by casesolved-co-uk** [comment]
I see this is blocked.
Congratulations 👏
More evidence you encourage bugs in frappe and erpnext

**Pull request closed without merging**